File: 2024/3/day3.js
const fs = require("fs");

const input = fs.readFileSync("./inputs/input.txt", "utf8");
// python regex: regex = r"mul\((\d{1,3}),(\d{1,3})\)"
const regex = /mul[\(\[]\d+,\d+[\)\]>](?!\(\d+,\d+\))/g;
const matches = input.matchAll(regex);

let sum = 0;
for (const match of matches) {
  sum += parseInt(match[1]) * parseInt(match[2]);
}

console.log(sum);

File: 2024/3/day3.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

func main() {
	input, _ := os.Open("./inputs/input.txt")
	defer input.Close()
	scanner := bufio.NewScanner(input)

	sum := 0
	regex := `mul\((\d{1,3}),(\d{1,3})\)`
	re := regexp.MustCompile(regex)

	// Read all lines from the file
	for scanner.Scan() {
		text := scanner.Text()
		matches := re.FindAllStringSubmatch(text, -1)

		for _, match := range matches {
			num1, _ := strconv.Atoi(match[1])
			num2, _ := strconv.Atoi(match[2])
			sum += num1 * num2
		}
	}

	fmt.Println(sum)
}
